# 342. Power of Four
### Problem:
    Given an integer (signed 32 bits), write a function to check whether it is a power of 4.

### This notebook shows the thought processes behind the solution to the problem.

In [1]:
# Every solution starts with the class, we shall copy that over:
class Solution(object):
    def isPowerOfFour(self, num):
        """
        :type num: int
        :rtype: bool
        """
        # Adding return as we'll be returning True or False
        return
 


Now we write a simple test class, it may be a nuisance, but it is good practice. 
Also once one is written, it can be pasted and the necessary changes can be easily made.

In [2]:
# To check if the values we return are correct import this.
from nose.tools import assert_equal

# class name taken from the function in the class above
class TestIsPowerOfFour(object):
    # function name also from the above class
    def test_isPowerOfFour(self):
        # define a varable to call the class where our solution is:
        solution = Solution()
        # use assert_equal(input, expected output)
        # Input: 
            # Call the function() within the class using '.', and the input
        # Output:
            # Expected output
        assert_equal(solution.isPowerOfFour(4), True)
        assert_equal(solution.isPowerOfFour(1), True)
        assert_equal(solution.isPowerOfFour(16), True)
        assert_equal(solution.isPowerOfFour(65536), True)
        assert_equal(solution.isPowerOfFour(0), False)
        assert_equal(solution.isPowerOfFour(-4), False)
        assert_equal(solution.isPowerOfFour(65535), False)
        assert_equal(solution.isPowerOfFour(123456), False)
        # An output to show tests have passed
        print('Success: test_isPowerOfFour')

# Call the tests
def main():
    test = TestIsPowerOfFour()
    test.test_isPowerOfFour()

# Run the tests when file is run
if __name__ == '__main__':
    main()
       
        
# Now with these placed in a script we can run it.
# The error below is expected.

AssertionError: None != True

Note the last line.
**AssertionError: None != True**
This means that one of our outputs isn't as expected.
Seeing as we haven't written code to solve the problem, it's hardly surprising!
Note the arrow pointing to line 15 above? That shows the output we have generated was not *True*.
Our output was *None*.

### Now let us get on with solving the problem

We have to find anything that a power of 4 would have in common.

Let us have a look at them:


In [3]:
# Using a list comprehension to generate powers of 4
print( [4**i for i in range(10)] )
# Printing the power
print( [i for i in range(10)] )


[1, 4, 16, 64, 256, 1024, 4096, 16384, 65536, 262144]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


So we have a sequence, where each subsequent value is 4 times bigger than the previous.
The relationship between the powers and the sequence can be found using logs.

$4^i = 1024$

$i \log (4) = \log (1024)$

$i = \frac{\log (1024)}{\log (4)}$

$i = 5$

So if we divide the log of a number by the log of 4 and we get an integer, then the number is a power of four!
Excellent, yes?
Well, *yes* but mostly *no*.

Let me show you why:


In [4]:
# Let's generate a larger power of 4
print(4**12)

16777216


In [5]:
# Let's do the log-ing as above:
from math import log
print(log(16777216)/log(4))

12.0


*Looks good doesn't it? a squeaky clean integer! This means $16777216$ is a power of four doesn't it?*

Yes.

*Right let us implement the code.*

No.

*Why? It works*

Let me show you this:


In [6]:
print(log(16777217)/log(4))

12.000000042995662


Can you see the issue? How can we define that 12.0 as an answer means a number ***is*** a power of four yet define that 12.000000042995662 means the number ***is not*** a power of four.

Well, we could generate an if statement to check whether a number is abritrarily accurate but that's a fudge which will fail at some point.

### There is another way:

Computers are *binary* machines and binary is incredibly handy for powers of 2, **four is a power of 2**.

It may be useful for us to check the binary representations of the powers of 4.

*(Note: It may be not be trivial to associate such powers with binary initially, but if you attempt to implement the log solution above and see how inelegant it is, this method of working soons become trivial)*

In [7]:
# Checking the output from using bin() on integers
[bin(i) for i in range(10)]

['0b0',
 '0b1',
 '0b10',
 '0b11',
 '0b100',
 '0b101',
 '0b110',
 '0b111',
 '0b1000',
 '0b1001']

In [8]:
# All outputs are strings which have a prefix 0b
# Let's output the binary strings for the first few powers of four

[bin(4**i) for i in range(10)]

['0b1',
 '0b100',
 '0b10000',
 '0b1000000',
 '0b100000000',
 '0b10000000000',
 '0b1000000000000',
 '0b100000000000000',
 '0b10000000000000000',
 '0b1000000000000000000']

Well, isn't that a pretty pattern!
let's remove the common prefix '0b' to get the actual binary number.
We shall save these to a variable.


In [9]:
bin_powers_of_4 = [bin(4**i)[2:] for i in range(10)]
bin_powers_of_4


['1',
 '100',
 '10000',
 '1000000',
 '100000000',
 '10000000000',
 '1000000000000',
 '100000000000000',
 '10000000000000000',
 '1000000000000000000']

Have these binary numbers have anything in common?

*They all have a single 1*

*The number of zeros is the sequence [0,2,4,....*

In [10]:
# count the number of '0's in each binary number in bin_powers_of_4
[number.count('0') for number in bin_powers_of_4]

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18]

We know that the number of ones is 1 and the number of zeros is an even number.

Is this the case only for powers of four? 

Let us try different values:


In [11]:
# Returning only the binary numbers with '0b' removed
# if they are not 1, 4, 16, i.e. powers of four
[bin(i)[2:] for i in range(20) if i not in [1,4,16]]

['0',
 '10',
 '11',
 '101',
 '110',
 '111',
 '1000',
 '1001',
 '1010',
 '1011',
 '1100',
 '1101',
 '1110',
 '1111',
 '10001',
 '10010',
 '10011']

We can see that other binary numbers do not fit our specific conditions.

Powers of 4 have ***a single '1'*** and ***an even number of zeros***

So, to check whether a number is a power of four:

1. Convert it to binary and strip the prefix

        bin(number)[0:2]
    
2. Check whether the number of 1s is one AND the number of zeros is an even number

        use count('0') == 1 and count('1') == even

How do we check for even?

Simple: If we divide a number by two, there is no remainder. In python we use the **%** operator.

In [12]:
# Printing a number and its remainder when divided by two.
print([(i, i % 2) for i in range(20)])

[(0, 0), (1, 1), (2, 0), (3, 1), (4, 0), (5, 1), (6, 0), (7, 1), (8, 0), (9, 1), (10, 0), (11, 1), (12, 0), (13, 1), (14, 0), (15, 1), (16, 0), (17, 1), (18, 0), (19, 1)]


Therefore, if the result from i % 2 is zero, the number is even.

It is time to implement a solution.

In [13]:
class Solution(object):
    def isPowerOfFour(self, num):
        """
        :type num: int
        :rtype: bool
        """
        # bin(num)[2:] is the binary number as a string
        # use .count on the string and our conditions
        if bin(num)[2:].count('1') == 1 and bin(num)[2:].count('0') % 2 == 0:
            # Condition met, is a power of four
            return True
        # Condition not met, is not a power of four
        return False

Let's run it, see which of our tests pass

In [14]:
main()

AssertionError: True != False

We have an error, we are returning *True* when it is actually *False*

The error code shows that it is from: *assert_equal(solution.isPowerOfFour(-4), False)*

Why is this? You can probably break your brain wondering if -4 is a power of four *(hint: it is not)*. But our code says that it is? Why?

In [15]:
bin(-4)

'-0b100'

it fits our conditions! After the prefix there is one *1* and an even number of *0s*

A little thought put into the test conditions has returned a valubale AssertionError.
Let's correct the code to remove values less than or equal to 0. 


In [16]:
class Solution(object):
    def isPowerOfFour(self, num):
        """
        :type num: int
        :rtype: bool
        """
        # Removing nums less than or equal to zero
        if num <= 0:
            return False
        # bin(num)[2:] is the binary number as a string
        # use .count on the string and our conditions
        if bin(num)[2:].count('1') == 1 and bin(num)[2:].count('0') % 2 == 0:
            # Condition met, is a power of four
            return True
        # Condition not met, is not a power of four
        return False

In [17]:
# And our tests:
main()

Success: test_isPowerOfFour


Fantabulous! All our tests pass, last thing to do is to see if the code can be tidied up.

As we are returning a boolean, true or false, we can condense the last few lines of code.

In [18]:
class Solution(object):
    def isPowerOfFour(self, num):
        """
        This function checks whether an inputted number is a power of four.
        :type num: int
        :rtype: bool
        
        Algorithm:  Filter out negatives
                    Check binary string for one, '1's and an even number of '0's
        """
        
        if num <= 0:
            return False
                
        return (bin(num)[2:].count('1') == 1 and bin(num)[2:].count('0') % 2 == 0)

In [19]:
# Checking we haven't broken anything
main()

Success: test_isPowerOfFour


---------------------

And there we have it, a function which successfully checks whether a passed integer is a power of four.

---------------------